web_basic/understand_cache/task_1/README.md

```diff
@@ -0,0 +1,227 @@
+  - [Macブラウザのキャッシュサイズ変更方法とキャッシュの場所](https://infornography.blue/mac/macos-browser-cache-size/)
+
+## 質問5
+
+> 先輩エンジニアに「動的なサイトをキャッシュするなら、expiresは使わない方が良いよ」と言われました。なぜでしょうか？どうすれば良いのでしょうか？
+
+### 回答
+
+- `Expires`を使用した期限切れモデルでのキャッシュ処理を行っていた場合、`Expires`の値の日付になるまで、サーバへの通信が行われずデータが更新されない。そのため、データが頻繁に変化する動的サイトのキャッシュには不向きである。
+
+## 質問6
+
+> ブラウザのキャッシュがWEBサービスに用いられている実例を、3つ以上見つけて共有してください。どのような仕組みで対象がキャッシュされているのか、技術的な流れを説明してみてください
+
+### 回答
+
+- [Qiita](https://qiita.com/)
+  - `etag: W/"2e709c47148dbf95592ae69eb859165a"`
+    - データに変更があるかをサーバに検証しにいく、検証モデルによるキャッシュの仕組みを導入していると考えられる。
+    - `W/`がついているため、弱い検証を利用している
+
+![](../../../assets/cache_qiita.png)
+
+- [PADI](https://www.padi.co.jp/)
+  - `Cache-Control: max-age=0`
+    - `max-age=0`のため、キャッシュは行わずにリクエストごとにサーバと通信していると考えられる。
+
+![](../../../assets/cache_padi.png)
+
+- [Amazon](https://www.amazon.co.jp/)
+  - `expires: -1`
+    - `Expires`の値は`HTTP-date timestamp`とされているので、正しい値の設定方法ではないように思われるが、詳細は不明...
+  - `pragma: no-cache`
+    - キャッシュするが、検証が必要
+  - `vary: Accept-Encoding,User-Agent,Content-Type,Accept-Encoding,X-Amzn-CDN-Cache,X-Amzn-AX-Treatment,User-Agent`
+    - リクエストヘッダの値によって、レスポンスデータが変化するために、`vary`ヘッダを使用していると考えられる。
+
+![](../../../assets/cache_amazon.png)
+
+---
+
+# キャッシュに関する調査内容
+
+## キャッシュについて
+
+- サーバから取得したリソースをローカルストレージ（ハードディスクなど）に蓄積し、再利用する手法
+- サーバへのアクセス頻度や通信量を減らすために使用される
+- キャッシュの種類に関しては、[質問2の回答](#質問-2)参照
+
+## キャッシュの方法
+
+### 期限切れモデル（Expiration Model）
+
+- 予めレスポンスデータに保存期限を決めておき、期限が切れたら再度アクセスをして取得を行う
+
+#### キャッシュ用のヘッダ
+
+|ヘッダ名|概要|値の例|備考|
+|--------|-----------------------------|---------|-----------|
+|`Pragma`|リソースをキャッシュするが検証が必要|[draft-ietf-httpbis-cache-latest](https://httpwg.org/http-core/draft-ietf-httpbis-cache-latest.html#field.pragma)では非推奨とされている。`no-cache`|HTTP/1.0クライアントとの後方互換用に限り使用する。それ以外の場合は`Cache-Control`ヘッダを使用|
+|`Expires`|キャッシュの有効期限を示す。絶対時間で指定する。|`Thu, 11 May 2010 16:00:00 GMT`|1年以上の日時を設定することは推奨されていない|
+|`Cache-Control`|キャッシュをするか否か、有効期限などのキャッシュに関する複雑な設定を行う。時間は相対時間を秒で指定する。|`no-cache`、`max-age: 86400`|`Pragma`ヘッダと`Expires`ヘッダの機能の上位互換とも言われるが、絶対時間を指定したい場合は`Expires`を使用する。`Expires`と`Cache-Control`の両方が指定されていた場合は、より新しい仕様の後者が優先される。<br> `max-age`の計算には、`Date`ヘッダの値が使用される。|
+
+- `no-cache`は、キャッシュしないわけではない！キャッシュするが、検証モデルによる検証が必要なことを表す（キャッシュしない場合は`no-store`を指定する）
+
+### 検証モデル（Validation Model）
+
+- 現在保持しているキャッシュが最新であるかを問い合わせて、データが更新されていた場合のみ取得を行う
+
+#### 条件付きGETに関わるヘッダ
+
+- キャッシュ用のヘッダを検証した結果、ローカルのキャッシュをそのまま利用できない場合でも、条件付きGETでキャッシュを再利用できる可能性がある
+- 以下のヘッダを保持している場合に条件付きGETを利用できる
+
+|ヘッダ名|概要|値の例|備考|
+|--------|-------|--------|----------|
+|`If-Modified-Since`|ローカルキャッシュの更新日時を示す。<br> サーバ上のリソースがこれ以降変更されていなければ、304 Not Modifiedで`Last-Modified`を返却する。|`Thu, 11 May 2010 16:00:00 GMT`|304のレスポンスにはボディが存在しないため、その分ネットワーク帯域を節約可能。|
+|`If-None-Match`|キャッシュしてあるリソースのETagヘッダの値。<br>サーバ上のリソースが変更されていなければ、304 Not ModifiedでEtagの値を返却する。|`ab3322028`|`ETag`はリソースの更新状態を比較するためだけに使用する文字列のため、リソースを更新した時に別の値になるのであればどのような文字列もOK。<br>|
```

Comment:
ここら辺のリクエストヘッダは、Cache-Controlにキャッシュを保存するディレクティブが設定されていた場合に、ブラウザが勝手にヘッダに追加するんですかね 🤔 

Chronium的にはキャッシュに保存されているレスポンスのヘッダにETagが含まれている場合に、`If-None-Match` を設定する感じでしょうか

https://source.chromium.org/chromium/chromium/src/+/master:content/browser/appcache/appcache_update_url_fetcher.cc;l=163-172?q=If-None-Match&ss=chromium

Reply:
これも実験してみたいですね。

Reply:
確認してみました。

`no-cache`を設定していた場合、
初回リクエストでは、キャッシュがまだないので、`If-Modified-Since`や`If-None-Match`ヘッダは存在しないですね。（当然か、、）
![image](https://user-images.githubusercontent.com/47489629/106915998-0986c480-674a-11eb-9e8d-c1e153c16468.png)

2回目のリクエストでは、`If-Modified-Since`や`If-None-Match`ヘッダがリクエストに追加されました。
ステータスコードは304を返しているので、おそらく`If-None-Match`ヘッダとレスポンスにある`ETag`ヘッダを用いた弱い検証が行われた結果、304を返したのかなと思っています。（アプリケーション側では`cache-control:no-cache`を設定する以外何もしていない）
`If-Modified-Since`ヘッダもありますが、[MDN](https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/If-Modified-Since)によると
> 本ヘッダーは If-None-Match との組み合わせで使用された場合、サーバーが If-None-Match に対応していない場合を除き無視されます。

とのことなので、無視されてETagによる検証が行われたのかなと推察してます。

![image](https://user-images.githubusercontent.com/47489629/106916099-26bb9300-674a-11eb-8fe4-4f6d45b6a5a9.png)

---
ちなみに`cache-control: no-store`を設定していても、`If-Modified-Since`、`If-None-Match`、`ETag`ヘッダは勝手についてますね。（サーバ側では`no-store`を設定する以外は何もしてません）
ですが、`no-cache`を設定している方と違って、毎回200 OKが返ってくるので、やはりキャッシュからは取得せずに必ずサーバから取得しているみたいです。

![image](https://user-images.githubusercontent.com/47489629/106915169-13f48e80-6749-11eb-8b6b-8699a9886edf.png)